File: answers/leetcode/Balanced Binary Tree/Balanced Binary Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def isBalanced(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """

        return self.getDepth(root) >= 0

    def getDepth(self, root):
        if root:
            left = self.getDepth(root.left)
            if left < 0:
                return left
            right = self.getDepth(root.right)
            if right < 0:
                return right
            if abs(left - right) > 1:
                return -1
            else:
                return max(left, right) + 1
        else:
            return 0

File: answers/leetcode/Balanced Binary Tree/Balanced Binary Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public boolean isBalanced(TreeNode root) {
        return this.checkHeight(root) >= 0;
    }

    private int checkHeight(TreeNode root) {
        if (null == root) {
            return 0;
        }
        int left = this.checkHeight(root.left);
        if (left < 0) {
            return -1;
        }
        int right = this.checkHeight(root.right);
        if (right < 0) {
            return -1;
        }
        if (Math.abs(left - right) > 1) {
            return -1;
        }
        return Math.max(left, right) + 1;
    }
}